File: ICPC/2023 Vietnam Nationals/f.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

signed main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr); cout.tie(nullptr);

    int T; cin >> T;
    while (T--) {
        string s; cin >> s;
        vector<array<int, 3>> pref = {{0,0,0}};
        int l = s.size();
        for (char& x : s) {
            auto [a, b, c] = pref.back();
            if (x == 'R') ++a;
            else if (x == 'P') ++b;
            else ++c;
            pref.push_back({a, b, c});
        }
        
        vector<pair<int,int>> ans;
        for (int k = 2; k < l+1; ++k) {
            int temp = 0;
            for (int pos = 0; pos < s.size(); pos += k) {
                int a = pref[min(pos+k, l)][0] - pref[pos][0];
                int b = pref[min(pos+k, l)][1] - pref[pos][1];
                int c = pref[min(pos+k, l)][2] - pref[pos][2];
                temp += max({a, b, c});
            }
            ans.push_back({temp, k});
        }
        sort(ans.rbegin(), ans.rend());
        cout << ans[0].second << endl;
    }

    return 0;
}

File: ICPC/2023 Vietnam Nationals/f.py
import sys
input = sys.stdin.readline
for _ in range(int(input())):
  s = input()
  pref = [(0,0,0) for _ in range(len(s)+1)]
  l = len(s)
  for i, x in enumerate(s):
    a,b,c = pref[i]
    if x == "R":
      a += 1
    elif x == "P":
      b += 1
    else:
      c += 1
    pref[i+1] = (a,b,c)
  
  ans = int(-1e9)
  best_k = 0
  for k in range(2, l + 1):
    temp = 0
    for pos in range(0, len(s),k):
      a = pref[min(pos+k,l)][0] - pref[pos][0]
      b = pref[min(pos+k,l)][1] - pref[pos][1]
      c = pref[min(pos+k,l)][2] - pref[pos][2]
      temp += max(a,b,c)
    if ans < temp:
      ans = temp
      best_k = k
  print(best_k)
    
